src/json/serialization.cr

```diff
@@ -202,11 +202,7 @@ module JSON
         # `%var`'s type must be exact to avoid type inference issues with
         # recursively defined serializable types
         {% for name, value in properties %}
-          %var{name} = {% if value[:has_default] || value[:nilable] %}
-                         nil.as(::Union(::Nil, {{value[:type]}}))
-                       {% else %}
-                         uninitialized ::Union({{value[:type]}})
-                       {% end %}
+          %var{name} = uninitialized ::Union({{value[:type]}})
```

Comment:
interesting why this condition was needed?

Reply:
The convoluted implementation of the finalizing steps (where the ivar is assigned) always assigned `@{{name}} = %var{name}` if `value[:nilable]`. So it needed a value and could not be uninitialized.

Reply:
With this patch, it only assigns if `%found{name}` so there's no need for a default initialization.

Reply:
Before https://github.com/crystal-lang/crystal/pull/13238 `%var{name}` was always initialized to `nil` and then that nil type was filtered out later when the ivar was not nilable. `uninitialized` is a much better tool, though. It doesn't add `Nil` to the type.